// supi/lib/install/getWantedDependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@pnpm/utils");
const guessPinnedVersionFromExistingSpec_1 = require("../guessPinnedVersionFromExistingSpec");
function getWantedDependencies(pkg, opts) {
    var _a, _b, _c;
    const depsToInstall = utils_1.filterDependenciesByType(pkg, (_b = (_a = opts) === null || _a === void 0 ? void 0 : _a.includeDirect, (_b !== null && _b !== void 0 ? _b : {
        dependencies: true,
        devDependencies: true,
        optionalDependencies: true,
    })));
    return getWantedDependenciesFromGivenSet(depsToInstall, {
        devDependencies: pkg.devDependencies || {},
        optionalDependencies: pkg.optionalDependencies || {},
        updatePref: ((_c = opts) === null || _c === void 0 ? void 0 : _c.updateWorkspaceDependencies) === true
            ? updateWorkspacePref
            : (pref) => pref,
    });
}
exports.default = getWantedDependencies;
function updateWorkspacePref(pref) {
    return pref.startsWith('workspace:') ? 'workspace:*' : pref;
}
function getWantedDependenciesFromGivenSet(deps, opts) {
    if (!deps)
        return [];
    return Object.keys(deps).map((alias) => {
        const pref = opts.updatePref(deps[alias]);
        return {
            alias,
            dev: !!opts.devDependencies[alias],
            optional: !!opts.optionalDependencies[alias],
            pinnedVersion: guessPinnedVersionFromExistingSpec_1.default(deps[alias]),
            pref,
            raw: `${alias}@${pref}`,
        };
    });
}
